File: backend/app/schemas/ai.py
"""Pydantic schemas matching the frontend Zod schemas in
`frontend/src/types/ai.ts`.

Field names are camelCase.  All models use ``extra = "forbid"`` to prevent
contract drift.  The anchor-id pattern mirrors the frontend anchorIdSchema.
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field

ANCHOR_ID_RE = r"^[a-z0-9_]+$"

Likelihood = Literal["low", "medium", "high", "unknown"]


# ---------------------------------------------------------------------------
# Request
# ---------------------------------------------------------------------------

class PatientInfo(BaseModel):
    age: int = Field(..., gt=0)
    sex: str = Field(..., min_length=1)
    smokingHistory: str = Field(..., min_length=1)

    model_config = {"extra": "forbid"}


class Tests(BaseModel):
    lungFunction: str | None = Field(default=None)
    ctDescription: str | None = Field(default=None)

    model_config = {"extra": "forbid"}


class AIAnalyzeRequest(BaseModel):
    patientInfo: PatientInfo
    symptoms: list[str] = Field(..., min_length=1)
    tests: Tests
    pathologyContext: list[str] = Field(default_factory=list)

    model_config = {"extra": "forbid"}


# ---------------------------------------------------------------------------
# Response
# ---------------------------------------------------------------------------

class AIAssessment(BaseModel):
    disease: str = Field(..., min_length=1)
    likelihood: Likelihood
    confidence: float = Field(..., ge=0, le=1)
    basis: list[str] = Field(..., min_length=1)

    model_config = {"extra": "forbid"}


class AIEvidence(BaseModel):
    text: str = Field(..., min_length=1)
    anchorId: str = Field(..., min_length=1, pattern=ANCHOR_ID_RE)

    model_config = {"extra": "forbid"}


class AIResult(BaseModel):
    assessment: AIAssessment
    evidence: list[AIEvidence] = Field(..., min_length=1)
    differential: list[str] = Field(default_factory=list)
    recommendation: list[str] = Field(default_factory=list)
    disclaimer: str = Field(..., min_length=1)

    model_config = {"extra": "forbid"}

File: frontend/src/types/ai.ts
import { z } from 'zod'

import { anchorIdSchema } from './anchor'

export const aiAssessmentSchema = z
  .object({
    disease: z.string().trim().min(1),
    likelihood: z.enum(['low', 'medium', 'high', 'unknown']),
    confidence: z.number().min(0).max(1),
    basis: z.array(z.string().trim().min(1)).min(1),
  })
  .strict()
export type AIAssessment = z.infer<typeof aiAssessmentSchema>

export const aiEvidenceSchema = z
  .object({
    text: z.string().trim().min(1),
    anchorId: anchorIdSchema,
  })
  .strict()
export type AIEvidence = z.infer<typeof aiEvidenceSchema>

export const aiResultSchema = z
  .object({
    assessment: aiAssessmentSchema,
    evidence: z.array(aiEvidenceSchema).min(1),
    differential: z.array(z.string().trim().min(1)),
    recommendation: z.array(z.string().trim().min(1)),
    disclaimer: z.string().trim().min(1),
  })
  .strict()
export type AIResult = z.infer<typeof aiResultSchema>

export const aiAnalyzeRequestSchema = z
  .object({
    patientInfo: z
      .object({
        age: z.number().int().positive(),
        sex: z.string().trim().min(1),
        smokingHistory: z.string().trim().min(1),
      })
      .strict(),
    symptoms: z.array(z.string().trim().min(1)).min(1),
    tests: z
      .object({
        lungFunction: z.string().trim().min(1).optional(),
        ctDescription: z.string().trim().min(1).optional(),
      })
      .strict(),
    pathologyContext: z.array(z.string().trim().min(1)),
  })
  .strict()
export type AIAnalyzeRequest = z.infer<typeof aiAnalyzeRequestSchema>
